ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +188,81 @@ public void shutdown (boolean async)
+     * Any attempt to start a new scan either by an ad hoc threads or by the periodic
+     * recovery thread will suspend its thread until the mode changes.
+     * If a scan is in progress when this method is called it will complete its scan
+     * without suspending.
+     * <p>Note that this method is also influenced by
+     * {@link com.arjuna.ats.arjuna.common.RecoveryEnvironmentBean#setWaitForRecovery}.
+     * In case <code>{@link RecoveryEnvironmentBean#isWaitForRecovery()} == true</code>,
+     * it is <b>very important</b> that, before invoking this method, the following
+     * pre-conditions are met:
+     * * The Transaction System gets disabled (i.e. {@link TxControl#disable()})
+     * (i.e. no new transactions will be created)
+     * * {@link com.arjuna.ats.arjuna.coordinator.TransactionReaper} finishes monitoring all in-flight
+     * transactions
+     * * All transactions without a timeout are beyond the prepare phase of the 2PC protocol;
+     * to check when this condition is verified, {@link com.arjuna.ats.arjuna.coordinator.ActionManager}
+     * can be employed
      *
-     * This switches the recovery operation mode to SUSPENDED. Any attempt to start a new scan either by an ad hoc
-     * threads or by the periodic recovery thread will suspend its thread until the mode changes. If a scan is in
-     * progress when this method is called it will complete its scan without suspending.
-     *
-     * @param async false if the calling thread should wait for any in-progress scan to complete before returning
-     * @return the previous mode before attempting the suspend
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     * In case <code>{@link RecoveryEnvironmentBean#isWaitForRecovery()} == true</code>, this parameter is
+     * override.
+     * @return the previous mode before attempting the suspension
      */
-
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForRecovery checks whether suspension should delay until RecoveryModule
+            * implementations recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForRecovery()) {
+
+               /*
+                * At least, one clear scan (after the Transaction Manager has been shut down)
+                * should be completed to make sure that there are no transactions to recover.
+                */
+               doScanningWait();
+               doWork();
+
+               /*
+                * Now, it is finally possible to start checking if there are transactions that
+                * are still in need of recovery (or completion, in case of heuristics).
```

Comment:
> or completion

A heuristic is in a final state so there is nothing to complete, they can only be resolved by clearing the heuristic. You need to ensure that the documentation is clear about how to resolve them otherwise we'll have a lots of unhappy customers complaining that Narayana is causing denial of service. What is your strategy for dealing with this scenario?

Also, I though the plan had been to provide sufficient information in the logs, or elsewhere, to enable users to resolve the heuristic without having to stall the system.

Also, some heuristics MAY not compromise the system, such as when an RM commits/aborts early and the TM's decision is to commit/abort and in such scenarios most users would be surprised that we block service just to let them know.

Reply:
> Also, I though the plan had been to provide sufficient information in the logs, or elsewhere, to enable users to resolve the heuristic without having to stall the system.

I think that maybe moving the list of heuristics to trace (instead of warn) could be considered for reverting so that it allows the user to see that list clearly.

Reply:
Here (https://github.com/jbosstm/documentation/pull/43/files https://github.com/jbosstm/documentation/pull/30/files) are some documentation pull request files where we have with documented aspects of dealing with heuristics at the moment but I expect we rely on having an Object Store.